Remove fingerprintjs fork, point to 3.x version directly

aduth opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #93475

## Proposed Changes

* Removes `fingerprintjs` package fork in favor of pointing directly to the original package, at a specific version in the 3.x line

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Improves install time by elimination a substantial number of dependencies (215), mostly installed via `devDependencies` of the package, which uses a different development / testing stack (Rollup, Karma, etc.)
* Removes unnecessary code. While the previous pull request didn't explicitly outline the reason for forking or changes made, the originating project thread makes references to [licensing changes in v4 of the package](https://fingerprint.com/blog/fingerprintjs-license-change/). An analysis of the differences between our package and v3.4.2 of the original package reveals that the two are effectively identical (37a7f-pb).
   * The only difference is the use of a `window.fingerprintJsVersion` global which (a) doesn't appear to be assigned in production environments, and (b) is used in adding a [monitoring request](https://github.com/fingerprintjs/fingerprintjs/blob/v3.4.2/src/agent.ts#L191) that we [disable](https://github.com/Automattic/wp-calypso/blob/f6fe7cbfaf9385d7d247ad09dac030b1106e9fd6/client/lib/wp/handleFrs/fingerprint.js#L5).

This does introduce a small risk that someone in the future upgrades to the latest version without recognizing the license change. Ideally someone would not make a major version upgrade without identifying the key differences in that major version, of which the license change is significantly noted [in the release notes](https://github.com/fingerprintjs/fingerprintjs/releases/tag/v4.0.0). Alternatively, we could consider additional tooling to either (a) assert the specific version of this specific package or (b) identify any dependencies with an incompatible license like BSL.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Verify that transaction requests continue to include the `X-Fingerprint` header:

1. Go to http://calypso.localhost:3000
2. Enable store sandbox
3. Open DevTools Network tab, optionally filtering to "transactions" to easily identify relevant request
4. Go to "Upgrades" and purchase an upgrade (use store sandbox workflow)
5. Look at "Headers" for `/me/transactions` request
6. Observe `X-Fingerprint` header with value of ~32 numbers and letters

Verify that added tests pass:

* `yarn test-client client/lib/wp/handlers/test/fingerprint.test.ts`
* Optional: Repeat on `trunk` with `git checkout remove/fingerprintjs-fork -- client/lib/wp/handlers/test/fingerprint.test.ts`

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aduth committed:
Add test coverage for fingerprint behavior

aduth committed:
Remove fingerprintjs fork, point to 3.x version directly

aduth committed:
Restore DefinitelyTyped types implicitly used through FingerprintJS

aduth commented on client/package.json:
```diff
@@ -235,6 +235,7 @@
 		"@testing-library/jest-dom": "^6.6.3",
 		"@testing-library/react": "^16.3.0",
 		"@testing-library/user-event": "^14.6.1",
+		"@types/express": "^5.0.1",
```
We were previously relying on this implicitly through the forked package, and it's no longer being installed. Since we reference Express directly ([source](https://github.com/Automattic/wp-calypso/blob/f2c8b08d24d6a893443d4cc61f06ab8663eaf611/client/server/api/index.ts#L2)), we should have it defined as an explicit dependency.

tyxla commented:
cc @Automattic/shilling could you folks confirm this is OK to remove? 

I've also asked in the Shilling team channel: p1746000871126859-slack-C0117V2PCAE

tyxla commented:
@aduth, in the meantime, we'll need to rebase, and it seems like we also should do a `yarn dedupe`.

tyxla reviewed (commented):
This looks great, thanks @aduth! 

Can we get a formal approval from @Automattic/shilling just in case (as per the relevant convo earlier: p1746000871126859-slack-C0117V2PCAE)?

jsnajdr commented on client/lib/wp/handlers/test/fingerprint.test.ts:
```diff
@@ -0,0 +1,80 @@
+
+		wpcom.request( { path: '/me/transactions' }, callback );
+
+		expect( originalRequest ).toHaveBeenCalledWith( { path: '/me/transactions' }, callback );
```
Seems to me that this test relies on a race condition between `DOMContentLoaded` and `wpcom.request`. The `loadFingerprint` function should be called also on `DOMContentLoaded`. Does JSDOM fire that event? If it does fire the event, what is the timing? If this test is to succeed, then it should be fired only after `wpcom.request` was already called.

aduth replied:
> Seems to me that this test relies on a race condition between `DOMContentLoaded` and `wpcom.request`. The `loadFingerprint` function should be called also on `DOMContentLoaded`. Does JSDOM fire that event? If it does fire the event, what is the timing? If this test is to succeed, then it should be fired only after `wpcom.request` was already called.

That's a great catch. After testing locally, it looks like `loadFingerprint` is being called as soon as the module is imported, but the `load` promise from `@fingerprintjs/fingerprintjs` isn't fully resolved by the time the "should not inject" test case is run. But this is very much prone to race conditions.

A few ideas for improving:

- Add a condition that only calls `loadFingerprint` in the module if `process.env.NODE_ENV` is not `'test'`
- Extract the DOM loading behavior to a separate module that's not loaded by the tests
- Stub the values of `document.readyState` and/or `document.addEventListener` to avoid calling `loadFingerprint`

aduth replied:
I addressed this in d8d706f203b. It's a little clunky to mock while ensuring load order of the module, but I preferred it against modifying the implementation to support the tests, and I think with the proposed revisions discussed in https://github.com/Automattic/wp-calypso/pull/102973#discussion_r2073843249 we may not need these tests for much longer.

jsnajdr commented on client/lib/wp/handlers/test/fingerprint.test.ts:
```diff
@@ -0,0 +1,80 @@
+			await loadFingerprint();
+		} );
+
+		test( 'should inject fingering header for transactions path', async () => {
```
typo: fingering -> fingerprint 🙂

aduth replied:
> typo: fingering -> fingerprint 🙂

🙃 Fixed in 3f4fcfc1964

jsnajdr commented on client/lib/wp/handlers/fingerprint.js:
```diff
@@ -1,7 +1,11 @@
+	if ( fingerprint ) {
+		return;
+	}
+
```
Some thoughts about the `lib/wp` handler itself:
- do we really need to add the `x-fingerprint` header here, at a centralized place? It's added to exactly one endpoint, `/me/transactions`, and that request is issued at exactly one place in Calypso, in the `submitWpcomTransaction` helper. What if we moved the fingerprinting there? Then it wouldn't need to be part of the centralized code that every entrypoint needs to load.
- if we decide to keep the `lib/wp` handler, the `loadFingerprint` call should be more "controlled". I.e., don't do it statically when the module is initialized, but do it only when the module exports are actually used. I.e., inside the `injectFingerprint` function. And a true lazy loading would be to load it inside the first actual request to `/me/transactions`.

aduth replied:
> Some thoughts about the `lib/wp` handler itself:
> 
> * do we really need to add the `x-fingerprint` header here, at a centralized place? It's added to exactly one endpoint, `/me/transactions`, and that request is issued at exactly one place in Calypso, in the `submitWpcomTransaction` helper. What if we moved the fingerprinting there? Then it wouldn't need to be part of the centralized code that every entrypoint needs to load.
> * if we decide to keep the `lib/wp` handler, the `loadFingerprint` call should be more "controlled". I.e., don't do it statically when the module is initialized, but do it only when the module exports are actually used. I.e., inside the `injectFingerprint` function. And a true lazy loading would be to load it inside the first actual request to `/me/transactions`.

These are good points. I had also wondered about whether the fingerprinting behavior could be deferred to be called-on-demand. I'd guessed the idea was to try to pre-load this so it's ready by the time a request is made, but it's unclear if that's a worthwhile trade-off to do the work upfront for any page load that loads this module, especially since almost everything uses `wpcom`.

I'm interested to take a look at moving the fingerprinting into places where `/me/transactions` is called, but may defer that for a follow-on pull request if it's extensive.

jsnajdr reviewed (commented):
This is a great find, I always assumed that this fork simply must be there because we're making some changes to it, and it turns out that we were just concerned that the old v3 version would be removed from NPM or that we would be loading it from some 3rd party URL at runtime.

aduth committed:
Fix race condition between fingerprint load and tests

aduth committed:
Remove duplicative loadFingerprint calls in tests

Was moved to shared beforeAll handler during iterations

aduth committed:
Remove load redundancy checks

Previously needed because both test and document load checking would separately initialize. Now we have stronger guarantees that loadFingerprint will be called at most once in both tests and in the browser

aduth committed:
Fix typo "fingerprint"

millerf reviewed (approved):
This looks good to me.
I cannot find any trace of why we did not go this path at first. But I will let FraudSquad know we made this change, so they can ping us if they see any weird thing happening with the fingerprints...
